### PR TITLE
Replace plugin endpoints with placeholder endpoints

### PR DESCRIPTION
**Pull request by Aiky30**
### Summary
Replace all plugin endpoints with the ones from the Placeholder admin

### Documentation checklist

* [ ] I have updated CHANGELOG.txt if appropriate
* [ ] I have updated the release notes document if appropriate, with:
    * [ ] general notes
    * [ ] bug-fixes
    * [ ] improvements/new features
    * [ ] backwards-incompatible changes
    * [ ] required upgrade steps
    * [ ] names of contributors
* [ ] I have updated other documentation
* [ ] I have added my name to the AUTHORS file
* [ ] This PR's documentation has been approved by Daniele Procida

### CODE REVIEW

File: cms/plugin_base.py [hunk 1]
```diff
@@ -333,6 +333,8 @@ def save_model(self, request, obj, form, change):
                 operation_kwargs['tree_order'] = tree_order
             # Remember the operation token
             self._operation_token = pl_admin._send_pre_placeholder_operation(**operation_kwargs)
+
```
Comment (czpython) [inline cms/plugin_base.py]: remove these

File: cms/tests/test_plugins.py [hunk 2]
```diff
@@ -1102,7 +1102,7 @@ def test_copy_plugin_with_m2m(self):
             language=self.FIRST_LANG,
         )
 
-        endpoint = self.get_admin_url(Page, 'edit_plugin', plugin.pk)
+        endpoint = self.get_admin_url(Placeholder, 'edit_plugin', plugin.pk)
```
Comment (czpython) [inline cms/tests/test_plugins.py]: replace any usage of get_admin_url for plugin endpoints with the respective get_<action>_plugin_uri helper method

File: cms/test_utils/testcases.py [hunk 3]
```diff
@@ -39,7 +39,8 @@
 from cms.utils.permissions import set_current_user
 from cms.utils.urlutils import admin_reverse
 
-
+# TODO: These should all be replaced using the models get url methods, see placeholder actions i.e. get_ACTION_url
```
Comment (czpython) [inline cms/test_utils/testcases.py]: remove

File: cms/tests/test_placeholder_admin.py [hunk 4]
```diff
@@ -26,15 +26,15 @@ def test_add_plugin_endpoint(self):
         self.assertEqual(response.status_code, 200)
         self.assertEqual(plugins.count(), 1)
 
-    def test_add_plugins_from_placeholder(self):
+    def test_copy_plugins_add_plugins_from_placeholder(self):
```
Comment (czpython) [inline cms/tests/test_placeholder_admin.py]: naming is not correct here, the name being removed is the correct one

File: cms/tests/test_placeholder_admin.py [hunk 5]
```diff
@@ -49,12 +49,12 @@ def test_add_plugins_from_placeholder(self):
         self.assertTrue(source_placeholder.get_plugins('en').filter(pk=source_plugin.pk).exists())
         self.assertTrue(
             target_placeholder
-            .get_plugins('en')
```
Comment (czpython) [inline cms/tests/test_placeholder_admin.py]: indentation is off

File: cms/tests/test_placeholder_admin.py [hunk 6]
```diff
@@ -49,12 +49,12 @@ def test_add_plugins_from_placeholder(self):
...
-            .filter(plugin_type=source_plugin.plugin_type)
-            .exists()
+                .get_plugins('en')
+                .filter(plugin_type=source_plugin.plugin_type)
+                .exists()
         )
 
-    def test_copy_plugins_to_clipboard(self):
+    def test_copy_plugins_copy_plugin_to_clipboard(self):
```
Comment (czpython) [inline cms/tests/test_placeholder_admin.py]: naming is not correct here, the name being removed is the correct one

File: cms/tests/test_placeholder_admin.py [hunk 7]
```diff
@@ -79,14 +79,15 @@ def test_copy_plugins_to_clipboard(self):
 
         # Test that the target placeholder has the plugin copied from the source placeholder (clipboard)
         self.assertEqual(response.status_code, 200)
+        self.assertTrue(source_placeholder.get_plugins('en').filter(pk=source_plugin.pk).exists())
         self.assertTrue(
             user_settings.clipboard
-            .get_plugins('en')
-            .filter(plugin_type=source_plugin.plugin_type)
-            .exists()
+                .get_plugins('en')
```
Comment (czpython) [inline cms/tests/test_placeholder_admin.py]: indentation is off

File: cms/tests/test_placeholder_admin.py [hunk 8]
```diff
@@ -79,14 +79,15 @@ def test_copy_plugins_to_clipboard(self):
...
+                .filter(plugin_type=source_plugin.plugin_type)
+                .exists()
         )
 
-    def test_copy_placeholder_to_clipboard(self):
+    def test_copy_plugins_copy_placeholder_to_clipboard(self):
```
Comment (czpython) [inline cms/tests/test_placeholder_admin.py]: naming is not correct here, the name being removed is the correct one

File: cms/tests/test_placeholder_admin.py [hunk 9]
```diff
@@ -112,9 +113,9 @@ def test_copy_placeholder_to_clipboard(self):
         self.assertTrue(source_placeholder.get_plugins('en').filter(pk=source_plugin.pk).exists())
         self.assertTrue(
             user_settings.clipboard
-            .get_plugins('en')
-            .filter(plugin_type='PlaceholderPlugin')
-            .exists()
+                .get_plugins('en')
```
Comment (czpython) [inline cms/tests/test_placeholder_admin.py]: indentation is off

File: cms/tests/test_placeholder_admin.py [hunk 10]
```diff
@@ -134,14 +135,42 @@ def test_edit_plugin_endpoint(self):
         self.assertEqual(response.status_code, 200)
         self.assertEqual(plugin.name, data['name'])
 
+    def test_move_plugin_endpoint(self):
```
Comment (czpython) [inline cms/tests/test_placeholder_admin.py]: remove this test

File: cms/tests/test_placeholder_app_admin.py [hunk 11]
```diff
@@ -699,6 +699,7 @@ def test_user_can_paste_from_clipboard(self):
         target_placeholder = self._obj.placeholder
 
         self.add_permission(staff_user, 'change_example1')
+        self.add_permission(staff_user, 'change_page')
```
Comment (czpython) [inline cms/tests/test_placeholder_app_admin.py]: why is this needed now?

File: cms/tests/test_plugins.py [hunk 12]
```diff
@@ -143,18 +141,16 @@ def _create_link_plugin_on_page(self, page, slot='col_left'):
...
 
+    def __edit_link_plugin(self, plugin, text):
+        endpoint = self.get_change_plugin_uri(plugin)
         response = self.client.get(endpoint)
         self.assertEqual(response.status_code, 200)
         data = {'name': text, 'external_link': 'https://www.django-cms.org'}
         response = self.client.post(endpoint, data)
         self.assertEqual(response.status_code, 200)
-        return CMSPlugin.objects.get(pk=plugin_id).get_bound_plugin()
+        return CMSPlugin.objects.get(pk=plugin.id).get_bound_plugin()
```
Comment (czpython) [inline cms/tests/test_plugins.py]: `plugin.pk`

File: cms/models/placeholdermodel.py [hunk 13]
```diff
@@ -109,16 +109,10 @@ def get_extra_menu_items(self):
         return plugin_pool.get_extra_placeholder_menu_items(self)
 
     def _get_url(self, key, pk=None):
```
Comment (czpython) [inline cms/models/placeholdermodel.py]: remove this method

File: cms/models/placeholdermodel.py [hunk 14]
```diff
@@ -289,15 +282,12 @@ def _get_attached_model(self):
 
     def _get_attached_admin(self, admin_site=None):
```
Comment (czpython) [inline cms/models/placeholdermodel.py]: remove this method

File: cms/test_utils/testcases.py [hunk 15]
```diff
@@ -529,85 +527,75 @@ def get_add_plugin_uri(self, placeholder, plugin_type, language='en', parent=Non
...
+    def get_change_plugin_uri(self, plugin, language=None):
         language = language or 'en'
 
         if plugin.page:
             path = plugin.page.get_absolute_url(language)
         else:
             path = '/{}/'.format(language)
 
-        endpoint = self.get_admin_url(container, 'edit_plugin', plugin.pk)
+        endpoint = plugin.placeholder.get_edit_url(plugin.pk)
```
Comment (czpython) [inline cms/test_utils/testcases.py]: change to use `admin_reverse`

File: cms/test_utils/testcases.py [hunk 16]
```diff
@@ -529,85 +527,75 @@ def get_add_plugin_uri(self, placeholder, plugin_type, language='en', parent=Non
...
+    def get_move_plugin_uri(self, plugin, language=None):
         language = language or 'en'
 
         if plugin.page:
             path = plugin.page.get_absolute_url(language)
         else:
             path = '/{}/'.format(language)
 
-        endpoint = self.get_admin_url(container, 'move_plugin')
+        endpoint = plugin.placeholder.get_move_url()
```
Comment (czpython) [inline cms/test_utils/testcases.py]: change to use `admin_reverse`

File: cms/test_utils/testcases.py [hunk 17]
```diff
@@ -529,85 +527,75 @@ def get_add_plugin_uri(self, placeholder, plugin_type, language='en', parent=Non
...
+    def get_copy_plugin_uri(self, plugin, language=None):
         language = language or 'en'
 
         if plugin.page:
             path = plugin.page.get_absolute_url(language)
         else:
             path = '/{}/'.format(language)
 
-        endpoint = self.get_admin_url(container, 'copy_plugins')
+        endpoint = plugin.placeholder.get_copy_url()
```
Comment (czpython) [inline cms/test_utils/testcases.py]: change to use `admin_reverse`

File: cms/test_utils/testcases.py [hunk 18]
```diff
@@ -529,85 +527,75 @@ def get_add_plugin_uri(self, placeholder, plugin_type, language='en', parent=Non
...
+    def get_copy_placeholder_uri(self, placeholder, language=None):
         language = language or 'en'
 
         if placeholder.page:
             path = placeholder.page.get_absolute_url(language)
         else:
             path = '/{}/'.format(language)
 
-        endpoint = self.get_admin_url(container, 'copy_plugins')
+        endpoint = placeholder.get_copy_url()
```
Comment (czpython) [inline cms/test_utils/testcases.py]: change to use `admin_reverse`

File: cms/test_utils/testcases.py [hunk 19]
```diff
@@ -529,85 +527,75 @@ def get_add_plugin_uri(self, placeholder, plugin_type, language='en', parent=Non
...
+    def get_delete_plugin_uri(self, plugin, language=None):
         language = language or 'en'
 
         if plugin.page:
             path = plugin.page.get_absolute_url(language)
         else:
             path = '/{}/'.format(language)
 
-        endpoint = self.get_admin_url(container, 'delete_plugin', plugin.pk)
+        endpoint = plugin.placeholder.get_delete_url(plugin.pk)
```
Comment (czpython) [inline cms/test_utils/testcases.py]: change to use `admin_reverse`

File: cms/test_utils/testcases.py [hunk 20]
```diff
@@ -529,85 +527,75 @@ def get_add_plugin_uri(self, placeholder, plugin_type, language='en', parent=Non
...
             path = placeholder.page.get_absolute_url(language)
         else:
             path = '/{}/'.format(language)
 
-        endpoint = self.get_admin_url(
-            container,
-            'clear_placeholder',
-            placeholder.pk,
-        )
+        endpoint = placeholder.get_clear_url()
```
Comment (czpython) [inline cms/test_utils/testcases.py]: change to use `admin_reverse`

File: cms/admin/placeholderadmin.py [hunk 21]
```diff
@@ -158,9 +158,51 @@ def edit_field(self, request, object_id, language):
 
 
 class PlaceholderAdminMixin(object):
+    class Meta:
```
[the remark this answers is not in the document]
Comment (Aiky30) [reply cms/admin/placeholderadmin.py]: You mentioned about adding a deprecation notice in the Meta class of the model  I couldn't find another example of where to put it. The __init__ method could be overridden by a class that extends it?

File: cms/admin/placeholderadmin.py [hunk 22]
```diff
@@ -158,9 +158,51 @@ def edit_field(self, request, object_id, language):
...
+    def has_change_permission(self, request, obj=None):
+        return False
+
+    def has_delete_permission(self, request, obj=None):
+        return False
+
+    def has_module_permission(self, request):
+        return False
+
+    def log_addition(self, request, object, object_repr):
```
[the remark this answers is not in the document]
Comment (Aiky30) [reply cms/admin/placeholderadmin.py]: Without the methods the admin creates log entries which causes the test_log_entries test case to fail because the placeholder operations are logged via signals and not the admin. They need to stay

File: cms/models/placeholdermodel.py (hunk 14 above)
Comment (Aiky30) [reply cms/models/placeholdermodel.py]: The save method of plugin base requires this:

    def save_model(self, request, obj, form, change):
        """
        Override original method, and add some attributes to obj
        This have to be made, because if object is newly created, he must know
        where he lives.
        """
        pl = obj.placeholder
        pl_admin = pl._get_attached_admin()

File: cms/tests/test_placeholder_admin.py (hunk 10 above)
Comment (Aiky30) [reply cms/tests/test_placeholder_admin.py]: Again I had already removed these. Merging the latest 4 to allow me to create a PR must have changed all of this back.
